system/jlib/jevent.cpp

```diff
@@ -838,17 +854,297 @@ EventRecorder eventRecorder;
+            case EvAttrEventThreadId:
+            case EvAttrEventStackTrace:
+                continue;
+            default:
+                if (definedAttributes.test(i) && !assignedAttributes.test(i))
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    virtual bool isTextAttribute(EventAttr attr) const override
+    {
+        assertex(attr < EvAttrMax);
+        return attributes[attr].isText();
+    }
+
+    virtual bool isNumericAttribute(EventAttr attr) const override
+    {
+        assertex(attr < EvAttrMax);
+        return attributes[attr].isNumeric();
+    }
+
+    virtual bool isBooleanAttribute(EventAttr attr) const override
+    {
+        assertex(attr < EvAttrMax);
+        return attributes[attr].isBoolean();
+    }
+
+    virtual const char* queryTextValue(EventAttr attr) const override
+    {
+        assertex(attr < EvAttrMax);
+        return attributes[attr].queryTextValue();
+    }
+
+    virtual __uint64 queryNumericValue(EventAttr attr) const override
+    {
+        assertex(attr < EvAttrMax);
+        return attributes[attr].queryNumericValue();
+    }
+
+    virtual bool queryBooleanValue(EventAttr attr) const override
+    {
+        assertex(attr < EvAttrMax);
+        return attributes[attr].queryBooleanValue();
+    }
+
+    virtual bool setValue(EventAttr attr, const char* value) override
+    {
+        assertex(attr < EvAttrMax);
+        if (attributes[attr].isText())
+        {
+            attributes[attr].setValue(value);
+            assignedAttributes.set(attr);
+            return true;
+        }
+        return false;
+    }
+
+    virtual bool setValue(EventAttr attr, __uint64 value) override
+    {
+        assertex(attr < EvAttrMax);
+        if (attributes[attr].isNumeric())
+        {
+            attributes[attr].setValue(value);
+            assignedAttributes.set(attr);
+            return true;
+        }
+        return false;
+    }
+
+    virtual bool setValue(EventAttr attr, bool value) override
+    {
+        assertex(attr < EvAttrMax);
+        if (attributes[attr].isBoolean())
+        {
+            attributes[attr].setValue(value);
+            assignedAttributes.set(attr);
+            return true;
+        }
+        return false;
+    }
+
+    virtual IEventAttributeIterator* getAttributes() const override
+    {
+        assertex(EventNone < type && type < EventMax);
+        return new AttributeIterator(*this);
+    }
+
+public:
+    CEvent()
+    {
+        for (unsigned i=0; i < EvAttrMax; i++)
+            attributes[i].setup(EventAttr(i));
+    }
+
+    void reset(EventType _type)
+    {
+        assertex(_type < EventMax);
+        type = _type;
+        orderedAttributes = (EventAttr*)eventInformation[type].attributes;
+        definedAttributes.reset();
+        for (unsigned i = EvAttrNone + 1; i < EvAttrMax; i++)
+            definedAttributes.set(i, true);
+        assignedAttributes.reset();
```

Comment:
Would this be better to include in attributes as a flag and avoid the extra bitset?